src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs

```diff
@@ -220,118 +220,58 @@ public override IEnumerable<FileSystemInfo> EnumerateFileSystemInfos(string full
-                findData = new Interop.Kernel32.WIN32_FIND_DATA();
-
-                // Remove trailing slash since this can cause grief to FindFirstFile. You will get an invalid argument error
-                string tempPath = path.TrimEnd(PathHelpers.DirectorySeparatorChars);
-
-                // For removable media drives, normally the OS will pop up a dialog requesting insertion
-                // of the relevant media (CD, floppy, memory card, etc.). We don't want this prompt so we
-                // set SEM_FAILCRITICALERRORS to suppress it.
-                //
-                // Note that said dialog only shows once the relevant filesystem has been loaded, which
-                // does not happen until actual media is accessed at least once since booting.
-
-                uint oldMode;
-                bool success = Interop.Kernel32.SetThreadErrorMode(Interop.Kernel32.SEM_FAILCRITICALERRORS, out oldMode);
-                try
+                if (!Interop.Kernel32.GetFileAttributesEx(path, Interop.Kernel32.GET_FILEEX_INFO_LEVELS.GetFileExInfoStandard, ref data))
                 {
-                    bool error = false;
-                    SafeFindHandle handle = Interop.Kernel32.FindFirstFile(tempPath, ref findData);
-                    try
+                    errorCode = Marshal.GetLastWin32Error();
+                    if (errorCode == Interop.Errors.ERROR_ACCESS_DENIED)
                     {
-                        if (handle.IsInvalid)
-                        {
-                            error = true;
-                            errorCode = Marshal.GetLastWin32Error();
+                        // Files that are marked for deletion will not let you GetFileAttributes,
+                        // ERROR_ACCESS_DENIED is given back without filling out the data struct.
+                        // FindFirstFile, however, will. Historically we always gave back attributes
+                        // for marked-for-deletion files.
 
-                            if (errorCode == Interop.Errors.ERROR_FILE_NOT_FOUND ||
-                                errorCode == Interop.Errors.ERROR_PATH_NOT_FOUND ||
-                                errorCode == Interop.Errors.ERROR_NOT_READY)  // Removable media not inserted
+                        var findData = new Interop.Kernel32.WIN32_FIND_DATA();
+                        using (SafeFindHandle handle = Interop.Kernel32.FindFirstFile(path, ref findData))
+                        {
+                            if (handle.IsInvalid)
                             {
-                                if (!returnErrorOnNotFound)
-                                {
-                                    // Return default value for backward compatibility
-                                    errorCode = 0;
-                                    data.fileAttributes = -1;
-                                }
+                                errorCode = Marshal.GetLastWin32Error();
```

Comment:
Do we still need this `returnErrorOnNotFound`?

If so, it looks like you're missing a check against it before setting the errorCode to something real here.